consensus/hotstuff/timeoutcollector/timeout_processor_test.go

```diff
@@ -0,0 +1,502 @@
+// timeout contains a proof that sender legitimately entered timeout.View but it has wrong view meaning he used TC from previous rounds.
+func (s *TimeoutProcessorTestSuite) TestProcess_LastViewTCWrongView() {
+	// if TC is included it must have timeout.View == timeout.LastViewTC.View+1
+	err := s.processor.Process(helper.TimeoutObjectFixture(
+		helper.WithTimeoutObjectView(s.view),
+		helper.WithTimeoutNewestQC(helper.MakeQC(helper.WithQCView(s.view-10))),
+		helper.WithTimeoutLastViewTC(helper.MakeTC(helper.WithTCView(s.view)))))
+	require.True(s.T(), model.IsInvalidTimeoutError(err))
+}
+
+// TestProcess_LastViewHighestQCInvalidView tests that TimeoutProcessor fails with model.InvalidTimeoutError if
+// timeout contains a proof that sender legitimately entered timeout.View but included HighestQC has older view
+// than QC included in TC. For honest nodes this shouldn't happen.
+func (s *TimeoutProcessorTestSuite) TestProcess_LastViewHighestQCInvalidView() {
+	err := s.processor.Process(helper.TimeoutObjectFixture(
+		helper.WithTimeoutObjectView(s.view),
+		helper.WithTimeoutNewestQC(helper.MakeQC(helper.WithQCView(s.view-10))),
+		helper.WithTimeoutLastViewTC(
+			helper.MakeTC(
+				helper.WithTCView(s.view-1),
+				helper.WithTCNewestQC(helper.MakeQC(helper.WithQCView(s.view-5)))))))
+	require.True(s.T(), model.IsInvalidTimeoutError(err))
+}
+
+// TestProcess_LastViewTCRequiredButNotPresent tests that TimeoutProcessor fails with model.InvalidTimeoutError if
+// timeout must contain a proof that sender legitimately entered timeout.View but doesn't have it.
+func (s *TimeoutProcessorTestSuite) TestProcess_LastViewTCRequiredButNotPresent() {
+	// if last view is not successful(timeout.View != timeout.HighestQC.View+1) then this
+	// timeout must contain valid timeout.LastViewTC
+	err := s.processor.Process(helper.TimeoutObjectFixture(
+		helper.WithTimeoutObjectView(s.view),
+		helper.WithTimeoutNewestQC(helper.MakeQC(helper.WithQCView(s.view-10))),
+		helper.WithTimeoutLastViewTC(nil)))
+	require.True(s.T(), model.IsInvalidTimeoutError(err))
+}
+
+// TestProcess_InvalidSigner tests that TimeoutProcessor correctly handles errors when timeout was created
+// by invalid signer.
+func (s *TimeoutProcessorTestSuite) TestProcess_InvalidSigner() {
+	timeout := helper.TimeoutObjectFixture(
+		helper.WithTimeoutObjectView(s.view),
+		helper.WithTimeoutNewestQC(helper.MakeQC(helper.WithQCView(s.view-1))),
+		helper.WithTimeoutLastViewTC(
+			helper.MakeTC(helper.WithTCView(s.view-1),
+				helper.WithTCNewestQC(helper.MakeQC(helper.WithQCView(s.view-1))))),
+	)
+
+	s.Run("invalid-signer", func() {
+		*s.committee = *mocks.NewReplicas(s.T())
+		s.committee.On("IdentityByEpoch", mock.Anything, mock.Anything).Return(nil, model.NewInvalidSignerError(errors.New(""))).Once()
+		err := s.processor.Process(timeout)
+		require.True(s.T(), model.IsInvalidTimeoutError(err))
+	})
```

Comment:
I'd suggest to add a test case that, if `IdentityByEpoch` returns `ErrViewForUnknownEpoch`, we should:
* expect an error from `Process`
* the error should not be of type `ErrViewForUnknownEpoch` or `InvalidTimeoutError` (ie. not an expected error)

Note: the second point is not reflected in the implementation yet, though I think we should update the implementation to be consistent with this test description (see [this comment](https://github.com/onflow/flow-go/pull/2620#discussion_r902017497))